File: app/src/main/java/me/ssttkkl/mrmemorizer/data/entity/Tree.java
package me.ssttkkl.mrmemorizer.data.entity;

import java.util.ArrayList;
import java.util.List;

import de.blox.graphview.Graph;
import de.blox.graphview.Node;

public class Tree {
    private String content;
    private List<Tree> children = new ArrayList<>();

    public Tree(String data){
        String[] strings = data.split("\n");
        this.content = strings[0];
        String child = "";
        for(int i=1;i<strings.length;i++){
            strings[i] = strings[i].substring(1);
            if (!strings[i].contains("\t")){
                if(!child.isEmpty()){
                    children.add(new Tree(child));
                }
                child = strings[i];
            }else{
                child += "\n"+strings[i];
            }
        }
        if(!child.isEmpty()){
            children.add(new Tree(child));
        }
    }

    public String getContent() {
        return content;
    }

    public List<Tree> getChildren() {
        return children;
    }

    public Graph convertToGraph(){
        Graph graph = new Graph();
        Node parentNode = new Node(getContent());
        graph.addNode(parentNode);
        for(Tree childTree:getChildren()){
            Node childNode = new Node(childTree.getContent());
            graph.addNode(childNode);
            graph.addEdge(parentNode,childNode);
            addSubTreeToGraph(graph,childNode,childTree);
        }
        return graph;
    }


    private void addSubTreeToGraph(Graph graph, Node parentNode, Tree tree){
        for(Tree childTree:tree.getChildren()){
            Node childNode = new Node(childTree.getContent());
            graph.addNode(childNode);
            graph.addEdge(parentNode,childNode);
            addSubTreeToGraph(graph,childNode,childTree);
        }
    }
}

File: app/src/main/java/me/ssttkkl/mrmemorizer/data/entity/Tree.kt
package me.ssttkkl.mrmemorizer.data.entity

import java.util.*

class Tree(
    val content: String,
    val children: Collection<Tree>
) {
    private fun appendDisplayText(sb: StringBuilder, prefixNum: Int) {
        for (i in 0 until prefixNum) sb.append(' ')
        sb.append(content)
        sb.append('\n')
        for (c in children) {
            c.appendDisplayText(sb, prefixNum + 1)
        }
    }

    fun toDisplayText(): String {
        val sb = StringBuilder()
        appendDisplayText(sb, 0)
        sb.deleteCharAt(sb.length - 1) // 删除最后一行的换行
        return sb.toString()
    }

    companion object {
        @JvmStatic
        fun parseText(text: String): Tree {
            val lines = text.split("\n").toMutableList()
            val root = lines[0]
            val prevChildText = StringBuilder()
            val children = ArrayList<Tree>()
            for (i in 1 until lines.size) {
                if (lines[i].isEmpty())
                    continue
                lines[i] = lines[i].substring(1)
                if (!lines[i].startsWith(" ")) {
                    if (prevChildText.isNotEmpty()) {
                        children.add(parseText(prevChildText.toString()))
                    }
                    prevChildText.delete(0, prevChildText.length)
                } else {
                    prevChildText.append('\n')
                }
                prevChildText.append(lines[i])
            }
            if (prevChildText.isNotEmpty()) {
                children.add(parseText(prevChildText.toString()))
            }
            return Tree(root, children)
        }
    }

}